src/darwin/Framework/CHIP/EndpointDescription/MTREndpointDescription.h

```diff
@@ -0,0 +1,70 @@
+/**
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+#import <Matter/MTRAccessGrant.h>
+#import <Matter/MTRDefines.h>
+#import <Matter/MTRDeviceType.h>
+#import <Matter/MTRServerClusterDescription.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * A representation of an endpoint implemented by an MTRDeviceController.
+ */
+MTR_NEWLY_AVAILABLE
+@interface MTREndpointDescription : NSObject <NSCopying>
+
+- (instancetype)init NS_UNAVAILABLE;
++ (instancetype)new NS_UNAVAILABLE;
+
+/**
+ * The provided endpointID must be in the range 1-65535.  The list of device
+ * types provided must be nonempty (but may include vendor-specific device
+ * types).
+ */
+- (nullable instancetype)initWithEndpointID:(NSNumber *)endpointID deviceTypes:(NSArray<MTRDeviceType *> *)deviceTypes error:(NSError * __autoreleasing *)error;
+
+@property (nonatomic, copy, readonly) NSNumber * endpointID;
+
+@property (nonatomic, copy, readonly) NSArray<MTRDeviceType *> * deviceTypes;
```

Comment:
Why is deviceTypes read only?

Reply:
Because it's (1) always required and (2) hence provided during init.

The writable bits are the ones which are basically optional, so don't have to be provided during init but can be set if desired.

I could also have a bunch of different init methods for all the combinations or something and have these objects always immutable, but that seemed like a bit of pain.  Any opinions on what's best here API-wise?